File: album-api-v2/src/models/album.ts
export interface Album {
  id: number;
  title: string;
  artist: string;
  price: number;
  image_url: string;
  year: number;
}

const seed: Album[] = [
  { id: 1, title: 'You, Me and an App Id', artist: 'Daprize', price: 10.99, image_url: 'https://aka.ms/albums-daprlogo', year: 2022 },
  { id: 2, title: 'Seven Revision Army', artist: 'The Blue-Green Stripes', price: 13.99, image_url: 'https://aka.ms/albums-containerappslogo', year: 2020 },
  { id: 3, title: 'Scale It Up', artist: 'KEDA Club', price: 13.99, image_url: 'https://aka.ms/albums-kedalogo', year: 2021 },
  { id: 4, title: 'Lost in Translation', artist: 'MegaDNS', price: 12.99, image_url: 'https://aka.ms/albums-envoylogo', year: 2019 },
  { id: 5, title: 'Lock Down Your Love', artist: 'V is for VNET', price: 12.99, image_url: 'https://aka.ms/albums-vnetlogo', year: 2023 },
  { id: 6, title: 'Sweet Container O\' Mine', artist: 'Guns N Probeses', price: 14.99, image_url: 'https://aka.ms/albums-containerappslogo', year: 2024 }
];

let albums: Album[] = seed.map(a => ({ ...a }));

export type AlbumInput = Omit<Album, 'id'>;

export const AlbumStore = {
  reset(): void {
    albums = seed.map(a => ({ ...a }));
  },

  getAll(): Album[] {
    return albums.slice();
  },

  getById(id: number): Album | undefined {
    return albums.find(a => a.id === id);
  },

  getByYear(year: number): Album[] {
    return albums.filter(a => a.year === year);
  },

  create(input: AlbumInput): Album {
    const nextId = albums.length === 0 ? 1 : Math.max(...albums.map(a => a.id)) + 1;
    const album: Album = { id: nextId, ...input };
    albums.push(album);
    return album;
  },

  update(id: number, input: AlbumInput): Album | null {
    const index = albums.findIndex(a => a.id === id);
    if (index < 0) return null;
    const updated: Album = { id, ...input };
    albums[index] = updated;
    return updated;
  },

  delete(id: number): boolean {
    const index = albums.findIndex(a => a.id === id);
    if (index < 0) return false;
    albums.splice(index, 1);
    return true;
  }
};

File: albums-api/Models/Album.cs
namespace albums_api.Models
{
    public record Album(int Id, string Title, string Artist, double Price, string Image_url, int Year)
    {
        private static readonly List<Album> Albums = new()
        {
            new Album(1, "You, Me and an App Id", "Daprize", 10.99, "https://aka.ms/albums-daprlogo", 2022),
            new Album(2, "Seven Revision Army", "The Blue-Green Stripes", 13.99, "https://aka.ms/albums-containerappslogo", 2020),
            new Album(3, "Scale It Up", "KEDA Club", 13.99, "https://aka.ms/albums-kedalogo", 2021),
            new Album(4, "Lost in Translation", "MegaDNS", 12.99, "https://aka.ms/albums-envoylogo", 2019),
            new Album(5, "Lock Down Your Love", "V is for VNET", 12.99, "https://aka.ms/albums-vnetlogo", 2023),
            new Album(6, "Sweet Container O' Mine", "Guns N Probeses", 14.99, "https://aka.ms/albums-containerappslogo", 2024)
        };

        public static List<Album> GetAll()
        {
            return Albums.ToList();
        }

        public static Album? GetById(int id)
        {
            return Albums.FirstOrDefault(a => a.Id == id);
        }

        public static List<Album> GetByYear(int year)
        {
            return Albums.Where(a => a.Year == year).ToList();
        }

        public static Album Create(string title, string artist, double price, string imageUrl, int year)
        {
            var nextId = Albums.Count == 0 ? 1 : Albums.Max(a => a.Id) + 1;
            var album = new Album(nextId, title, artist, price, imageUrl, year);
            Albums.Add(album);

            return album;
        }

        public static Album? Update(int id, string title, string artist, double price, string imageUrl, int year)
        {
            var index = Albums.FindIndex(a => a.Id == id);
            if (index < 0)
            {
                return null;
            }

            var updatedAlbum = new Album(id, title, artist, price, imageUrl, year);
            Albums[index] = updatedAlbum;

            return updatedAlbum;
        }

        public static bool Delete(int id)
        {
            var album = Albums.FirstOrDefault(a => a.Id == id);
            if (album is null)
            {
                return false;
            }

            return Albums.Remove(album);
        }
    }
}
